src/main/java/hudson/plugins/build_timeout/BuildStepWithTimeout.java

```diff
@@ -0,0 +1,153 @@
+package hudson.plugins.build_timeout;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.model.*;
+import hudson.plugins.build_timeout.impl.AbsoluteTimeOutStrategy;
+import hudson.plugins.build_timeout.impl.DeadlineTimeOutStrategy;
+import hudson.plugins.build_timeout.operations.AbortOperation;
+import hudson.tasks.BuildStep;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.BuildStepMonitor;
+import hudson.tasks.Builder;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.tokenmacro.MacroEvaluationException;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.StaplerRequest;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+public class BuildStepWithTimeout extends Builder implements BuildStep {
+    private /* final */ BuildTimeOutStrategy strategy;
+    private BuildStep buildStep;
+    private final List<BuildTimeOutOperation> operationList;
+
+    @DataBoundConstructor
+    public BuildStepWithTimeout(BuildStep buildStep, BuildTimeOutStrategy strategy, List<BuildTimeOutOperation> operationList) {
+        this.strategy = strategy;
+        this.buildStep = buildStep;
+        if (operationList != null && !operationList.isEmpty()) {
+            this.operationList = operationList;
+        }
+        else {
+            this.operationList = Collections.emptyList();
+        }
+
+    }
+
+    public List<BuildTimeOutOperation> getOperationList() {
+        return operationList;
+    }
+
+
+    public BuildStep getBuildStep() {
+        return buildStep;
+    }
+
+    public void setBuildStep(BuildStep buildStep) {
+        this.buildStep = buildStep;
+    }
+
+    private long getTimeout(Run run, TaskListener listener) throws IOException, InterruptedException {
+        try {
+            return strategy.getTimeOut((AbstractBuild<?, ?>) run, (BuildListener) listener);
+        } catch (MacroEvaluationException e) {
+            e.printStackTrace(listener.getLogger());
+            listener.error("Can't evaluate timeout - timeout would be disabled");
+            return Long.MAX_VALUE;
+        }
+    }
+
+
+    public boolean perform(final Build<?,?> build, final Launcher launcher, final BuildListener listener) throws InterruptedException, IOException {
+        final Timer timer = new Timer("Timer-" + build.getDisplayName().replace(" ", "_"), true);
```

Comment:
You can use `Trigger.timer` as in `BuildTimeoutWrapper`.
http://javadoc.jenkins-ci.org/hudson/triggers/Trigger.html#timer